<center> <font size = 6> Census Income (KDD) </font> </center>

Load the **train data** into Pandas dataframe.

In [1]:
import pandas as pd
df=pd.read_csv("KDD_census_income.csv",header=None)
df.head()

0                                1   2   3                            4   \
0  73                  Not in universe   0   0         High school graduate   
1  58   Self-employed-not incorporated   4  34   Some college but no degree   
2  18                  Not in universe   0   0                   10th grade   
3   9                  Not in universe   0   0                     Children   
4  10                  Not in universe   0   0                     Children   

   5                 6               7                             8   \
0   0   Not in universe         Widowed   Not in universe or children   
1   0   Not in universe        Divorced                  Construction   
2   0       High school   Never married   Not in universe or children   
3   0   Not in universe   Never married   Not in universe or children   
4   0   Not in universe   Never married   Not in universe or children   

                                     9   ...              32              33  \
0                       Not in universe  ...   United-States   United-States   
1   Precision production craft & repair  ...   United-States   United-States   
2                       Not in universe  ...         Vietnam         Vietnam   
3                       Not in universe  ...   United-States   United-States   
4                       Not in universe  ...   United-States   United-States   

               34                                    35 36                37  \
0   United-States     Native- Born in the United States  0   Not in universe   
1   United-States     Native- Born in the United States  0   Not in universe   
2         Vietnam   Foreign born- Not a citizen of U S   0   Not in universe   
3   United-States     Native- Born in the United States  0   Not in universe   
4   United-States     Native- Born in the United States  0   Not in universe   

   38  39  40      41  
0   2   0  95  -50000  
1   2  52  94  -50000  
2   2   0  95  -50000  
3   0   0  94  -50000  
4   0   0  94  -50000  

[5 rows x 42 columns]

In [2]:
df.shape

(199523, 42)

The train data has 199,523 rows and 42 columns.

Replace all missing categorical values that are currently denoted by "?" with the word "unknown".

In [3]:
df=df.replace(" ?", "unknown")
df

0                                1   2   3   \
0       73                  Not in universe   0   0   
1       58   Self-employed-not incorporated   4  34   
2       18                  Not in universe   0   0   
3        9                  Not in universe   0   0   
4       10                  Not in universe   0   0   
...     ..                              ...  ..  ..   
199518  87                  Not in universe   0   0   
199519  65       Self-employed-incorporated  37   2   
199520  47                  Not in universe   0   0   
199521  16                  Not in universe   0   0   
199522  32                          Private  42  30   

                                 4   5                 6   \
0              High school graduate   0   Not in universe   
1        Some college but no degree   0   Not in universe   
2                        10th grade   0       High school   
3                          Children   0   Not in universe   
4                          Children   0   Not in universe   
...                             ...  ..               ...   
199518            7th and 8th grade   0   Not in universe   
199519                   11th grade   0   Not in universe   
199520   Some college but no degree   0   Not in universe   
199521                   10th grade   0       High school   
199522         High school graduate   0   Not in universe   

                                      7                              8   \
0                                Widowed    Not in universe or children   
1                               Divorced                   Construction   
2                          Never married    Not in universe or children   
3                          Never married    Not in universe or children   
4                          Never married    Not in universe or children   
...                                  ...                            ...   
199518   Married-civilian spouse present    Not in universe or children   
199519   Married-civilian spouse present   Business and repair services   
199520   Married-civilian spouse present    Not in universe or children   
199521                     Never married    Not in universe or children   
199522                     Never married        Medical except hospital   

                                          9   ...              32  \
0                            Not in universe  ...   United-States   
1        Precision production craft & repair  ...   United-States   
2                            Not in universe  ...         Vietnam   
3                            Not in universe  ...   United-States   
4                            Not in universe  ...   United-States   
...                                      ...  ...             ...   
199518                       Not in universe  ...          Canada   
199519        Executive admin and managerial  ...   United-States   
199520                       Not in universe  ...          Poland   
199521                       Not in universe  ...   United-States   
199522                         Other service  ...         unknown   

                    33              34  \
0        United-States   United-States   
1        United-States   United-States   
2              Vietnam         Vietnam   
3        United-States   United-States   
4        United-States   United-States   
...                ...             ...   
199518   United-States   United-States   
199519   United-States   United-States   
199520          Poland         Germany   
199521   United-States   United-States   
199522         unknown         unknown   

                                                  35 36                37  38  \
0                  Native- Born in the United States  0   Not in universe   2   
1                  Native- Born in the United States  0   Not in universe   2   
2                Foreign born- Not a citizen of U S   0   Not in universe   2   
3                  Native- Born in the United States  0

A lot of the time, missing numeric values will are denoted by large numbers such as 9999, 99999, etc. They are usually the largest numbers in any numeric column. We don't really need the summary statistics, but the "describe" method can reveal the columns with such values.

In [4]:
df.describe()

0              2              3              5   \
count  199523.000000  199523.000000  199523.000000  199523.000000   
mean       34.494199      15.352320      11.306556      55.426908   
std        22.310895      18.067129      14.454204     274.896454   
min         0.000000       0.000000       0.000000       0.000000   
25%        15.000000       0.000000       0.000000       0.000000   
50%        33.000000       0.000000       0.000000       0.000000   
75%        50.000000      33.000000      26.000000       0.000000   
max        90.000000      51.000000      46.000000    9999.000000   

                 16             17             18             24  \
count  199523.00000  199523.000000  199523.000000  199523.000000   
mean      434.71899      37.313788     197.529533    1740.380269   
std      4697.53128     271.896428    1984.163658     993.768156   
min         0.00000       0.000000       0.000000      37.870000   
25%         0.00000       0.000000       0.000000    1061.615000   
50%         0.00000       0.000000       0.000000    1618.310000   
75%         0.00000       0.000000       0.000000    2188.610000   
max     99999.00000    4608.000000   99999.000000   18656.300000   

                  30             36             38             39  \
count  199523.000000  199523.000000  199523.000000  199523.000000   
mean        1.956180       0.175438       1.514833      23.174897   
std         2.365126       0.553694       0.851473      24.411488   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       2.000000       0.000000   
50%         1.000000       0.000000       2.000000       8.000000   
75%         4.000000       0.000000       2.000000      52.000000   
max         6.000000       2.000000       2.000000      52.000000   

                  40  
count  199523.000000  
mean       94.499672  
std         0.500001  
min        94.000000  
25%        94.000000  
50%        94.000000  
75%        95.000000  
max        95.000000

Columns 5, 16,18 have missing values and the missing values are being represented by large numbers such as 9999, 99999.

In [5]:
df_5=pd.DataFrame(df.iloc[:,5])
df_5[df_5.loc[:,5]==9999]
mean_5=df_5.drop(index=190936,axis=0).mean()
df_5=df_5.replace(9999,mean_5)

In [6]:
df_16=pd.DataFrame(df.iloc[:,16])
missing=df_16[df_16.loc[:,16]==99999]
index=missing.index
mean_16=df_16.drop(index=index,axis=0).mean()
df_16=df_16.replace(99999,mean_16)

In [7]:
df_18=pd.DataFrame(df.iloc[:,18])
missing=df_18[df_18.loc[:,18]==99999]
index=missing.index
mean_18=df_18.drop(index=index,axis=0).mean()
df_18=df_18.replace(99999,mean_18)

The cleaned dataset:

In [8]:
import numpy as np
df=df.drop(columns=[5,16,18],axis=1)
df=pd.concat([df_5,df_16,df_18,df],axis=1)
columns=np.arange(0,42)
df=df.reindex(columns=columns)
df.head()

0                                1   2   3                            4   \
0  73                  Not in universe   0   0         High school graduate   
1  58   Self-employed-not incorporated   4  34   Some college but no degree   
2  18                  Not in universe   0   0                   10th grade   
3   9                  Not in universe   0   0                     Children   
4  10                  Not in universe   0   0                     Children   

    5                 6               7                             8   \
0  0.0   Not in universe         Widowed   Not in universe or children   
1  0.0   Not in universe        Divorced                  Construction   
2  0.0       High school   Never married   Not in universe or children   
3  0.0   Not in universe   Never married   Not in universe or children   
4  0.0   Not in universe   Never married   Not in universe or children   

                                     9   ...              32              33  \
0                       Not in universe  ...   United-States   United-States   
1   Precision production craft & repair  ...   United-States   United-States   
2                       Not in universe  ...         Vietnam         Vietnam   
3                       Not in universe  ...   United-States   United-States   
4                       Not in universe  ...   United-States   United-States   

               34                                    35 36                37  \
0   United-States     Native- Born in the United States  0   Not in universe   
1   United-States     Native- Born in the United States  0   Not in universe   
2         Vietnam   Foreign born- Not a citizen of U S   0   Not in universe   
3   United-States     Native- Born in the United States  0   Not in universe   
4   United-States     Native- Born in the United States  0   Not in universe   

   38  39  40      41  
0   2   0  95  -50000  
1   2  52  94  -50000  
2   2   0  95  -50000  
3   0   0  94  -50000  
4   0   0  94  -50000  

[5 rows x 42 columns]

In [9]:
X=df.drop(columns=41,axis=1)
X.head()

0                                1   2   3                            4   \
0  73                  Not in universe   0   0         High school graduate   
1  58   Self-employed-not incorporated   4  34   Some college but no degree   
2  18                  Not in universe   0   0                   10th grade   
3   9                  Not in universe   0   0                     Children   
4  10                  Not in universe   0   0                     Children   

    5                 6               7                             8   \
0  0.0   Not in universe         Widowed   Not in universe or children   
1  0.0   Not in universe        Divorced                  Construction   
2  0.0       High school   Never married   Not in universe or children   
3  0.0   Not in universe   Never married   Not in universe or children   
4  0.0   Not in universe   Never married   Not in universe or children   

                                     9   ...                     31  \
0                       Not in universe  ...        Not in universe   
1   Precision production craft & repair  ...        Not in universe   
2                       Not in universe  ...        Not in universe   
3                       Not in universe  ...   Both parents present   
4                       Not in universe  ...   Both parents present   

               32              33              34  \
0   United-States   United-States   United-States   
1   United-States   United-States   United-States   
2         Vietnam         Vietnam         Vietnam   
3   United-States   United-States   United-States   
4   United-States   United-States   United-States   

                                     35 36                37  38  39  40  
0     Native- Born in the United States  0   Not in universe   2   0  95  
1     Native- Born in the United States  0   Not in universe   2  52  94  
2   Foreign born- Not a citizen of U S   0   Not in universe   2   0  95  
3     Native- Born in the United States  0   Not in universe   0   0  94  
4     Native- Born in the United States  0   Not in universe   0   0  94  

[5 rows x 41 columns]

Use the "describe" method to get the column indices of the numeric columns.

In [10]:
X.describe().columns

Int64Index([0, 2, 3, 5, 16, 17, 18, 24, 30, 36, 38, 39, 40], dtype='int64')

Separate the numeric columns from the categorical columns.

In [11]:
X_numeric=X.iloc[:,[0, 2, 3, 5, 16, 17, 18, 24, 30, 36, 38, 39, 40]]
X_numeric

0   2   3    5       16  17     18       24  30  36  38  39  40
0       73   0   0  0.0     0.0   0    0.0  1700.09   0   0   2   0  95
1       58   4  34  0.0     0.0   0    0.0  1053.55   1   0   2  52  94
2       18   0   0  0.0     0.0   0    0.0   991.95   0   0   2   0  95
3        9   0   0  0.0     0.0   0    0.0  1758.14   0   0   0   0  94
4       10   0   0  0.0     0.0   0    0.0  1069.16   0   0   0   0  94
...     ..  ..  ..  ...     ...  ..    ...      ...  ..  ..  ..  ..  ..
199518  87   0   0  0.0     0.0   0    0.0   955.27   0   0   2   0  95
199519  65  37   2  0.0  6418.0   0    9.0   687.19   1   0   2  52  94
199520  47   0   0  0.0     0.0   0  157.0  1923.03   6   0   2  52  95
199521  16   0   0  0.0     0.0   0    0.0  4664.87   0   0   2   0  95
199522  32  42  30  0.0     0.0   0    0.0  1830.11   6   0   2  52  94

[199523 rows x 13 columns]

In [12]:
X_categorical=X.drop(columns=[0, 2, 3, 5, 16, 17, 18, 24, 30, 36, 38, 39, 40],axis=1)

In [13]:
X_categorical

1                            4   \
0                       Not in universe         High school graduate   
1        Self-employed-not incorporated   Some college but no degree   
2                       Not in universe                   10th grade   
3                       Not in universe                     Children   
4                       Not in universe                     Children   
...                                 ...                          ...   
199518                  Not in universe            7th and 8th grade   
199519       Self-employed-incorporated                   11th grade   
199520                  Not in universe   Some college but no degree   
199521                  Not in universe                   10th grade   
199522                          Private         High school graduate   

                      6                                 7   \
0        Not in universe                           Widowed   
1        Not in universe                          Divorced   
2            High school                     Never married   
3        Not in universe                     Never married   
4        Not in universe                     Never married   
...                  ...                               ...   
199518   Not in universe   Married-civilian spouse present   
199519   Not in universe   Married-civilian spouse present   
199520   Not in universe   Married-civilian spouse present   
199521       High school                     Never married   
199522   Not in universe                     Never married   

                                   8                                     9   \
0         Not in universe or children                       Not in universe   
1                        Construction   Precision production craft & repair   
2         Not in universe or children                       Not in universe   
3         Not in universe or children                       Not in universe   
4         Not in universe or children                       Not in universe   
...                               ...                                   ...   
199518    Not in universe or children                       Not in universe   
199519   Business and repair services        Executive admin and managerial   
199520    Not in universe or children                       Not in universe   
199521    Not in universe or children                       Not in universe   
199522        Medical except hospital                         Other service   

                                10          11       12                13  \
0                            White   All other   Female   Not in universe   
1                            White   All other     Male   Not in universe   
2        Asian or Pacific Islander   All other   Female   Not in universe   
3                            White   All other   Female   Not in universe   
4                            White   All other   Female   Not in universe   
...                            ...         ...      ...               ...   
199518                       White   All other     Male   Not in universe   
199519                       White   All other     Male   Not in universe   
199520                       White   All other     Male   Not in universe   
199521                       White   All other   Female   Not in universe   
199522                       Black   All other   Female                No   

        ...            26            27                                 28  \
0       ...       unknown       unknown   Not in universe under 1 year old   
1       ...   Same county   Same county                                 No   
2       ...       unknown       unknown   Not in universe under 1 year old   
3       ...      Nonmover      Nonmover                                Yes   
4       ...      Nonmover      Nonmover                                Yes   
...     ...           ...           ...                                ...   
199518 

Label encode the categorical variables. 

In [14]:
from sklearn import preprocessing
le=preprocessing.LabelEncoder()
X_categorical=X_categorical.apply(le.fit_transform)
X_categorical

1   4   6   7   8   9   10  11  12  13  ...  26  27  28  29  31  32  \
0        3  12   2   6  14   6   4   0   0   1  ...   8   9   1   3   4  39   
1        6  16   2   0   4   8   4   0   1   1  ...   7   8   0   2   4  39   
2        3   0   1   4  14   6   1   0   0   1  ...   8   9   1   3   4  40   
3        3  10   2   4  14   6   4   0   0   1  ...   5   6   2   1   0  39   
4        3  10   2   4  14   6   4   0   0   1  ...   5   6   2   1   0  39   
...     ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..   
199518   3   5   2   2  14   6   4   0   1   1  ...   8   9   1   3   4   1   
199519   5   1   2   2   2   2   4   0   1   1  ...   5   6   2   1   4  39   
199520   3  16   2   2  14   6   4   0   1   1  ...   8   9   1   3   4  31   
199521   3   0   1   4  14   6   4   0   0   1  ...   8   9   1   3   0  39   
199522   4  12   2   4  12   7   2   0   0   0  ...   5   6   2   1   4  42   

        33  34  35  37  
0       39  39   4   1  
1       39  39   4   1  
2       40  40   0   1  
3       39  39   4   1  
4       39  39   4   1  
...     ..  ..  ..  ..  
199518  39  39   4   1  
199519  39  39   4   1  
199520  31  10   1   1  
199521  39  39   4   1  
199522  42  42   0   1  

[199523 rows x 28 columns]

Concatenate the encoded categorical variables and the numeric variables together.

In [15]:
X=pd.concat([X_numeric,X_categorical],axis=1)
X.head()

0   2   3    5    16  17   18       24  30  36  ...  26  27  28  29  31  \
0  73   0   0  0.0  0.0   0  0.0  1700.09   0   0  ...   8   9   1   3   4   
1  58   4  34  0.0  0.0   0  0.0  1053.55   1   0  ...   7   8   0   2   4   
2  18   0   0  0.0  0.0   0  0.0   991.95   0   0  ...   8   9   1   3   4   
3   9   0   0  0.0  0.0   0  0.0  1758.14   0   0  ...   5   6   2   1   0   
4  10   0   0  0.0  0.0   0  0.0  1069.16   0   0  ...   5   6   2   1   0   

   32  33  34  35  37  
0  39  39  39   4   1  
1  39  39  39   4   1  
2  40  40  40   0   1  
3  39  39  39   4   1  
4  39  39  39   4   1  

[5 rows x 41 columns]

Standardize the predictor variables.

In [16]:
from sklearn import preprocessing
names=X.columns
scaler=preprocessing.StandardScaler()
X=scaler.fit_transform(X)
X=pd.DataFrame(X, columns=names)
X.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


0         2         3         5         16        17        18  \
0  1.725879 -0.849740 -0.782235 -0.202111 -0.147203 -0.137236 -0.112838   
1  1.053560 -0.628343  1.570028 -0.202111 -0.147203 -0.137236 -0.112838   
2 -0.739291 -0.849740 -0.782235 -0.202111 -0.147203 -0.137236 -0.112838   
3 -1.142682 -0.849740 -0.782235 -0.202111 -0.147203 -0.137236 -0.112838   
4 -1.097861 -0.849740 -0.782235 -0.202111 -0.147203 -0.137236 -0.112838   

         24        30        36  ...        26        27       28        29  \
0 -0.040543 -0.827096 -0.316851  ...  0.896424  0.870045 -0.54229  0.965129   
1 -0.691139 -0.404284 -0.316851  ...  0.294051  0.294521 -2.16289  0.020509   
2 -0.753126 -0.827096 -0.316851  ...  0.896424  0.870045 -0.54229  0.965129   
3  0.017871 -0.827096 -0.316851  ... -0.910695 -0.856525  1.07831 -0.924110   
4 -0.675431 -0.827096 -0.316851  ... -0.910695 -0.856525  1.07831 -0.924110   

         31        32        33        34        35        37  
0  0.589025  0.367189  0.363905  0.268967  0.343561  0.060524  
1  0.589025  0.367189  0.363905  0.268967  0.343561  0.060524  
2  0.589025  0.487063  0.484627  0.421523 -3.254776  0.060524  
3 -1.901181  0.367189  0.363905  0.268967  0.343561  0.060524  
4 -1.901181  0.367189  0.363905  0.268967  0.343561  0.060524  

[5 rows x 41 columns]

Get the target variable.

In [17]:
y=pd.DataFrame(df.loc[:,41]).rename(columns={41:"income"})
y.head()

income
0  -50000
1  -50000
2  -50000
3  -50000
4  -50000

Feature importance scores for all features.

In [18]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.feature_selection import SelectFromModel
clf = ExtraTreesClassifier(n_estimators=50)
clf = clf.fit(X, y)
scores=clf.feature_importances_ 
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  after removing the cwd from sys.path.


scores
2   0.090260
0   0.084387
6   0.075489
7   0.072644
4   0.065359
14  0.062001
11  0.047337
1   0.047086
8   0.043428
17  0.038006
21  0.037376
18  0.034777
13  0.031809
5   0.021459
28  0.020235
25  0.019411
16  0.018604
22  0.017876
9   0.017218
36  0.016323
37  0.015022
29  0.014186
19  0.012246
24  0.011379
3   0.011013
20  0.009102
39  0.008038
38  0.007911
32  0.005386
31  0.005370
34  0.005283
30  0.005254
33  0.005086
10  0.004320
12  0.004197
27  0.003714
23  0.003200
26  0.003165
40  0.002037
35  0.001643
15  0.001359

Select the top 20 features.

In [19]:
top_20=scores.nlargest(20,"scores")
top_20.index

Int64Index([2, 0, 6, 7, 4, 14, 11, 1, 8, 17, 21, 18, 13, 5, 28, 25, 16, 22, 9,
            36],
           dtype='int64')

Form the predicto set with the top 20 features.

In [20]:
X=X.iloc[:,top_20.index]
X.head()

3         0         18        24        16        4         39  \
0 -0.782235  1.725879 -0.112838 -0.040543 -0.147203  0.476524 -0.949346   
1  1.570028  1.053560 -0.112838 -0.691139 -0.147203  1.447186  1.180804   
2 -0.782235 -0.739291 -0.112838 -0.753126 -0.147203 -2.435461 -0.949346   
3 -0.782235 -1.142682 -0.112838  0.017871 -0.147203 -0.008807 -0.949346   
4 -0.782235 -1.097861 -0.112838 -0.675431 -0.147203 -0.008807 -0.949346   

         2         30        8         12        9         1         17  \
0 -0.849740 -0.827096  0.207994 -0.958533 -0.097301 -0.439146 -0.137236   
1 -0.628343 -0.404284 -1.884785  1.043261  0.546914  2.275109 -0.137236   
2 -0.849740 -0.827096  0.207994 -0.958533 -0.097301 -0.439146 -0.137236   
3 -0.849740 -0.827096  0.207994 -0.958533 -0.097301 -0.439146 -0.137236   
4 -0.849740 -0.827096  0.207994 -0.958533 -0.097301 -0.439146 -0.137236   

         22        19        7         13        36        32  
0  0.344851  0.564160  2.126157  0.215699 -0.316851  0.367189  
1 -0.005216 -2.323698 -2.134072  0.215699 -0.316851  0.367189  
2 -1.580516  0.564160  0.706081  0.215699 -0.316851  0.487063  
3 -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189  
4 -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189

In [21]:
df=pd.concat([X,y],axis=1)
df

3         0        18        24        16         4        39  \
0      -0.782235  1.725879 -0.112838 -0.040543 -0.147203  0.476524 -0.949346   
1       1.570028  1.053560 -0.112838 -0.691139 -0.147203  1.447186  1.180804   
2      -0.782235 -0.739291 -0.112838 -0.753126 -0.147203 -2.435461 -0.949346   
3      -0.782235 -1.142682 -0.112838  0.017871 -0.147203 -0.008807 -0.949346   
4      -0.782235 -1.097861 -0.112838 -0.675431 -0.147203 -0.008807 -0.949346   
...          ...       ...       ...       ...       ...       ...       ...   
199518 -0.782235  2.353376 -0.112838 -0.790036 -0.147203 -1.222134 -0.949346   
199519 -0.643867  1.367309 -0.107349 -1.059797  3.793791 -2.192796  1.180804   
199520 -0.782235  0.560526 -0.017090  0.183796 -0.147203  1.447186  1.180804   
199521 -0.782235 -0.828933 -0.112838  2.942836 -0.147203 -2.435461 -0.949346   
199522  1.293291 -0.111793 -0.112838  0.090293 -0.147203  0.476524  1.180804   

               2        30         8  ...         9         1        17  \
0      -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
1      -0.628343 -0.404284 -1.884785  ...  0.546914  2.275109 -0.137236   
2      -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
3      -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
4      -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
...          ...       ...       ...  ...       ...       ...       ...   
199518 -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
199519  1.198183 -0.404284 -2.303341  ... -1.385731  1.370357 -0.137236   
199520 -0.849740  1.709774  0.207994  ... -0.097301 -0.439146 -0.137236   
199521 -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
199522  1.474930  1.709774 -0.210562  ...  0.224807  0.465605 -0.137236   

              22        19         7        13        36        32  income  
0       0.344851  0.564160  2.126157  0.215699 -0.316851  0.367189  -50000  
1      -0.005216 -2.323698 -2.134072  0.215699 -0.316851  0.367189  -50000  
2      -1.580516  0.564160  0.706081  0.215699 -0.316851  0.487063  -50000  
3      -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189  -50000  
4      -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189  -50000  
...          ...       ...       ...       ...       ...       ...     ...  
199518 -0.005216 -1.601733 -0.713995  0.215699 -0.316851 -4.188020  -50000  
199519 -0.005216 -0.157804 -0.713995  0.215699 -0.316851  0.367189  -50000  
199520 -0.005216 -0.879769 -0.713995  0.215699 -0.316851 -0.591802  -50000  
199521 -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189  -50000  
199522  0.169818  1.286125  0.706081 -3.093817 -0.316851  0.726811  -50000  

[199523 rows x 21 columns]

Check whether classes are evenly distributed.

In [22]:
y.groupby(["income"])["income"].count()

income
 50000+.     12382
-50000      187141
Name: income, dtype: int64

Un-even class distribution. Re-calibration is needed. We are going to over-sample the minority class "50000+."

In [23]:
df_above=df[df["income"]==" 50000+."]
df_above

3         0        18        24        16         4        39  \
57     -0.643867  0.336419 -0.112838 -0.796224 -0.147203  1.447186  1.180804   
64      1.500844  0.560526 -0.112838 -0.468461 -0.147203 -1.950130  1.180804   
68      1.500844 -0.066972  5.985731  0.124576 -0.147203 -0.494138  1.180804   
80      0.393896  1.053560 -0.112838  0.244222 -0.147203  1.204521  1.180804   
88     -0.782235  0.874275 -0.112838  1.307370 -0.147203  0.476524 -0.949346   
...          ...       ...       ...       ...       ...       ...       ...   
199474 -0.643867  0.605347 -0.112838 -0.654843 -0.147203 -0.251472  1.098875   
199487 -0.643867  0.963917 -0.112838 -0.188838 -0.147203 -0.251472  1.180804   
199499 -0.643867  1.098381 -0.112838 -0.207303 -0.147203 -0.251472  1.180804   
199508 -0.643867  1.232845 -0.112838  0.036135 -0.147203  1.447186  1.180804   
199509  0.047975  0.560526 -0.112838  0.890944 -0.147203  0.233859  1.180804   

               2        30         8  ...         9         1        17  \
57      1.198183  0.018527 -2.303341  ... -1.385731  1.370357 -0.137236   
64      0.866088  1.709774  1.882218  ...  0.546914  0.465605 -0.137236   
68      0.810738  1.709774 -2.094063  ...  0.546914  0.465605  8.851504   
80      1.087485 -0.404284 -1.256952  ...  1.835344  2.275109 -0.137236   
88     -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
...          ...       ...       ...  ...       ...       ...       ...   
199474  1.087485 -0.404284 -1.256952  ... -1.385731  2.275109 -0.137236   
199487  0.866088  1.709774  1.882218  ... -1.385731  0.465605 -0.137236   
199499  0.478643  0.018527 -0.419840  ... -1.385731  0.465605 -0.137236   
199508 -0.628343  1.709774 -1.884785  ... -1.385731  0.465605 -0.137236   
199509  1.530279  1.709774 -1.675507  ...  1.191129 -2.248650 -0.137236   

              22        19         7        13        36        32    income  
57     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
64     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
68      0.169818  1.286125  1.416119  0.215699 -0.316851  0.367189   50000+.  
80     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
88     -0.005216  0.564160 -0.713995  0.215699 -0.316851  0.487063   50000+.  
...          ...       ...       ...       ...       ...       ...       ...  
199474  0.169818  1.286125 -0.003957  0.215699 -0.316851 -1.670667   50000+.  
199487 -0.005216 -0.879769 -0.713995 -3.093817 -0.316851  0.726811   50000+.  
199499  1.482568 -0.157804 -0.713995  0.215699 -0.316851  0.367189   50000+.  
199508  0.169818  1.286125 -0.003957  0.215699  3.295258 -2.989280   50000+.  
199509  1.482568 -0.879769 -0.713995 -3.093817  3.295258  0.367189   50000+.  

[12382 rows x 21 columns]

In [24]:
df_above=pd.concat([df_above]*14,axis=0)
df_above

3         0        18        24        16         4        39  \
57     -0.643867  0.336419 -0.112838 -0.796224 -0.147203  1.447186  1.180804   
64      1.500844  0.560526 -0.112838 -0.468461 -0.147203 -1.950130  1.180804   
68      1.500844 -0.066972  5.985731  0.124576 -0.147203 -0.494138  1.180804   
80      0.393896  1.053560 -0.112838  0.244222 -0.147203  1.204521  1.180804   
88     -0.782235  0.874275 -0.112838  1.307370 -0.147203  0.476524 -0.949346   
...          ...       ...       ...       ...       ...       ...       ...   
199474 -0.643867  0.605347 -0.112838 -0.654843 -0.147203 -0.251472  1.098875   
199487 -0.643867  0.963917 -0.112838 -0.188838 -0.147203 -0.251472  1.180804   
199499 -0.643867  1.098381 -0.112838 -0.207303 -0.147203 -0.251472  1.180804   
199508 -0.643867  1.232845 -0.112838  0.036135 -0.147203  1.447186  1.180804   
199509  0.047975  0.560526 -0.112838  0.890944 -0.147203  0.233859  1.180804   

               2        30         8  ...         9         1        17  \
57      1.198183  0.018527 -2.303341  ... -1.385731  1.370357 -0.137236   
64      0.866088  1.709774  1.882218  ...  0.546914  0.465605 -0.137236   
68      0.810738  1.709774 -2.094063  ...  0.546914  0.465605  8.851504   
80      1.087485 -0.404284 -1.256952  ...  1.835344  2.275109 -0.137236   
88     -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
...          ...       ...       ...  ...       ...       ...       ...   
199474  1.087485 -0.404284 -1.256952  ... -1.385731  2.275109 -0.137236   
199487  0.866088  1.709774  1.882218  ... -1.385731  0.465605 -0.137236   
199499  0.478643  0.018527 -0.419840  ... -1.385731  0.465605 -0.137236   
199508 -0.628343  1.709774 -1.884785  ... -1.385731  0.465605 -0.137236   
199509  1.530279  1.709774 -1.675507  ...  1.191129 -2.248650 -0.137236   

              22        19         7        13        36        32    income  
57     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
64     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
68      0.169818  1.286125  1.416119  0.215699 -0.316851  0.367189   50000+.  
80     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
88     -0.005216  0.564160 -0.713995  0.215699 -0.316851  0.487063   50000+.  
...          ...       ...       ...       ...       ...       ...       ...  
199474  0.169818  1.286125 -0.003957  0.215699 -0.316851 -1.670667   50000+.  
199487 -0.005216 -0.879769 -0.713995 -3.093817 -0.316851  0.726811   50000+.  
199499  1.482568 -0.157804 -0.713995  0.215699 -0.316851  0.367189   50000+.  
199508  0.169818  1.286125 -0.003957  0.215699  3.295258 -2.989280   50000+.  
199509  1.482568 -0.879769 -0.713995 -3.093817  3.295258  0.367189   50000+.  

[173348 rows x 21 columns]

Concatenate the majority class and the over-sampled minority class together.

In [25]:
df_below=df[df["income"]=="-50000"]
df=pd.concat([df_above,df_below],axis=0)
df

3         0        18        24        16         4        39  \
57     -0.643867  0.336419 -0.112838 -0.796224 -0.147203  1.447186  1.180804   
64      1.500844  0.560526 -0.112838 -0.468461 -0.147203 -1.950130  1.180804   
68      1.500844 -0.066972  5.985731  0.124576 -0.147203 -0.494138  1.180804   
80      0.393896  1.053560 -0.112838  0.244222 -0.147203  1.204521  1.180804   
88     -0.782235  0.874275 -0.112838  1.307370 -0.147203  0.476524 -0.949346   
...          ...       ...       ...       ...       ...       ...       ...   
199518 -0.782235  2.353376 -0.112838 -0.790036 -0.147203 -1.222134 -0.949346   
199519 -0.643867  1.367309 -0.107349 -1.059797  3.793791 -2.192796  1.180804   
199520 -0.782235  0.560526 -0.017090  0.183796 -0.147203  1.447186  1.180804   
199521 -0.782235 -0.828933 -0.112838  2.942836 -0.147203 -2.435461 -0.949346   
199522  1.293291 -0.111793 -0.112838  0.090293 -0.147203  0.476524  1.180804   

               2        30         8  ...         9         1        17  \
57      1.198183  0.018527 -2.303341  ... -1.385731  1.370357 -0.137236   
64      0.866088  1.709774  1.882218  ...  0.546914  0.465605 -0.137236   
68      0.810738  1.709774 -2.094063  ...  0.546914  0.465605  8.851504   
80      1.087485 -0.404284 -1.256952  ...  1.835344  2.275109 -0.137236   
88     -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
...          ...       ...       ...  ...       ...       ...       ...   
199518 -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
199519  1.198183 -0.404284 -2.303341  ... -1.385731  1.370357 -0.137236   
199520 -0.849740  1.709774  0.207994  ... -0.097301 -0.439146 -0.137236   
199521 -0.849740 -0.827096  0.207994  ... -0.097301 -0.439146 -0.137236   
199522  1.474930  1.709774 -0.210562  ...  0.224807  0.465605 -0.137236   

              22        19         7        13        36        32    income  
57     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
64     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
68      0.169818  1.286125  1.416119  0.215699 -0.316851  0.367189   50000+.  
80     -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189   50000+.  
88     -0.005216  0.564160 -0.713995  0.215699 -0.316851  0.487063   50000+.  
...          ...       ...       ...       ...       ...       ...       ...  
199518 -0.005216 -1.601733 -0.713995  0.215699 -0.316851 -4.188020    -50000  
199519 -0.005216 -0.157804 -0.713995  0.215699 -0.316851  0.367189    -50000  
199520 -0.005216 -0.879769 -0.713995  0.215699 -0.316851 -0.591802    -50000  
199521 -1.055416  0.564160  0.706081  0.215699 -0.316851  0.367189    -50000  
199522  0.169818  1.286125  0.706081 -3.093817 -0.316851  0.726811    -50000  

[360489 rows x 21 columns]

In [26]:
X=df.drop(["income"],axis=1)
X.head()

3         0         18        24        16        4         39  \
57 -0.643867  0.336419 -0.112838 -0.796224 -0.147203  1.447186  1.180804   
64  1.500844  0.560526 -0.112838 -0.468461 -0.147203 -1.950130  1.180804   
68  1.500844 -0.066972  5.985731  0.124576 -0.147203 -0.494138  1.180804   
80  0.393896  1.053560 -0.112838  0.244222 -0.147203  1.204521  1.180804   
88 -0.782235  0.874275 -0.112838  1.307370 -0.147203  0.476524 -0.949346   

          2         30        8         12        9         1         17  \
57  1.198183  0.018527 -2.303341  1.043261 -1.385731  1.370357 -0.137236   
64  0.866088  1.709774  1.882218  1.043261  0.546914  0.465605 -0.137236   
68  0.810738  1.709774 -2.094063  1.043261  0.546914  0.465605  8.851504   
80  1.087485 -0.404284 -1.256952 -0.958533  1.835344  2.275109 -0.137236   
88 -0.849740 -0.827096  0.207994  1.043261 -0.097301 -0.439146 -0.137236   

          22        19        7         13        36        32  
57 -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189  
64 -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189  
68  0.169818  1.286125  1.416119  0.215699 -0.316851  0.367189  
80 -0.005216 -0.879769 -0.713995  0.215699 -0.316851  0.367189  
88 -0.005216  0.564160 -0.713995  0.215699 -0.316851  0.487063

In [27]:
y=df[["income"]]
y.head()

income
57   50000+.
64   50000+.
68   50000+.
80   50000+.
88   50000+.

Build the Random Forest model.

In [28]:
from sklearn.ensemble import RandomForestClassifier
Tree = RandomForestClassifier(n_estimators=750,max_depth=34, random_state=0,criterion="entropy",max_features="sqrt")
Tree

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=34, max_features='sqrt', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=750, n_jobs=None,
            oob_score=False, random_state=0, verbose=0, warm_start=False)

In [29]:
Tree.fit(X,y)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=34, max_features='sqrt', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=750, n_jobs=None,
            oob_score=False, random_state=0, verbose=0, warm_start=False)

In [30]:
from sklearn.metrics import accuracy_score
y_pred=Tree.predict(X)
print("Accuracy on train data: ",accuracy_score(y,y_pred))

Accuracy on train data:  0.9992759834558058


Load **KDD's test data**.

In [31]:
df_test=pd.read_csv("KDD_census_income_test.csv",header=None)
df_test

0                                1   2   3   \
0      38                          Private   6  36   
1      44   Self-employed-not incorporated  37  12   
2       2                  Not in universe   0   0   
3      35                          Private  29   3   
4      49                          Private   4  34   
...    ..                              ...  ..  ..   
99757  14                  Not in universe   0   0   
99758  61                          Private   8  36   
99759  24   Self-employed-not incorporated   1  43   
99760  30                          Private  45   2   
99761  67                  Not in universe   0   0   

                                         4   5                 6   \
0                  1st 2nd 3rd or 4th grade   0   Not in universe   
1       Associates degree-occup /vocational   0   Not in universe   
2                                  Children   0   Not in universe   
3                      High school graduate   0   Not in universe   
4                      High school graduate   0   Not in universe   
...                                     ...  ..               ...   
99757                              Children   0   Not in universe   
99758                            11th grade   0   Not in universe   
99759                     7th and 8th grade   0   Not in universe   
99760            Bachelors degree(BA AB BS)   0   Not in universe   
99761                             9th grade   0   Not in universe   

                                     7                              8   \
0       Married-civilian spouse present    Manufacturing-durable goods   
1       Married-civilian spouse present   Business and repair services   
2                         Never married    Not in universe or children   
3                              Divorced                 Transportation   
4                              Divorced                   Construction   
...                                 ...                            ...   
99757                     Never married    Not in universe or children   
99758                         Separated    Manufacturing-durable goods   
99759   Married-civilian spouse present                    Agriculture   
99760   Married-civilian spouse present    Other professional services   
99761   Married-civilian spouse present    Not in universe or children   

                                           9   ...              32  \
0       Machine operators assmblrs & inspctrs  ...          Mexico   
1                      Professional specialty  ...   United-States   
2                             Not in universe  ...   United-States   
3              Executive admin and managerial  ...   United-States   
4         Precision production craft & repair  ...   United-States   
...                                       ...  ...             ...   
99757                         Not in universe  ...   United-States   
99758   Machine operators assmblrs & inspctrs  ...   United-States   
99759            Farming forestry and fishing  ...          Mexico   
99760          Executive admin and managerial  ...   United-States   
99761                         Not in universe  ...   United-States   

                   33              34  \
0              Mexico          Mexico   
1       United-States   United-States   
2       United-States   United-States   
3       United-States   United-States   
4       United-States   United-States   
...               ...             ...   
99757   United-States   United-States   
99758   United-States   United-States   
99759          Mexico          Mexico   
99760   United-States   United-States   
99761   United-States   United-States   

                                                 35 36                37  38  \
0               Foreign born- Not a citizen of U S   0   Not in universe   2   
1                 Native- Born in the United States  0   Not in universe   2   
2                 Native- Born in the United States  0   Not in un

Clean the test data the way the train data was cleaned.

In [32]:
df_test=df_test.replace(" ?", "unknown")

In [33]:
df_test.describe()

0             2             3             5             16  \
count  99762.000000  99762.000000  99762.000000  99762.000000  99762.000000   
mean      34.628596     15.332812     11.331118     54.461268    425.788577   
std       22.333449     18.026709     14.459589    270.202643   4616.795578   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.000000      0.000000      0.000000      0.000000      0.000000   
50%       33.000000      0.000000      0.000000      0.000000      0.000000   
75%       50.000000     33.000000     26.000000      0.000000      0.000000   
max       90.000000     51.000000     46.000000   9900.000000  99999.000000   

                 17            18            24            30            36  \
count  99762.000000  99762.000000  99762.000000  99762.000000  99762.000000   
mean      35.919458    192.494727   1739.542838      1.956156      0.178505   
std      265.525252   1841.728492    994.900902      2.364331      0.558316   
min        0.000000      0.000000     43.260000      0.000000      0.000000   
25%        0.000000      0.000000   1058.050000      0.000000      0.000000   
50%        0.000000      0.000000   1615.395000      1.000000      0.000000   
75%        0.000000      0.000000   2187.277500      4.000000      0.000000   
max     4608.000000  99999.000000  16258.200000      6.000000      2.000000   

                 38            39            40  
count  99762.000000  99762.000000  99762.000000  
mean       1.519827     23.185331     94.500652  
std        0.848229     24.389184      0.500002  
min        0.000000      0.000000     94.000000  
25%        2.000000      0.000000     94.000000  
50%        2.000000      8.000000     95.000000  
75%        2.000000     52.000000     95.000000  
max        2.000000     52.000000     95.000000

In [34]:
df_test_16=pd.DataFrame(df_test.iloc[:,16])
missing=df_test_16[df_test_16.loc[:,16]==99999]
index=missing.index
mean_16=df_test_16.drop(index=index,axis=0).mean()
df_test_16=df_test_16.replace(99999,mean_16)

In [35]:
df_test_18=pd.DataFrame(df_test.iloc[:,18])
missing=df_test_18[df_test_18.loc[:,18]==99999]
index=missing.index
mean_18=df_test_18.drop(index=index,axis=0).mean()
df_test_18=df_test_18.replace(99999,mean_18)

In [36]:
import numpy as np
df_test=df_test.drop(columns=[16,18],axis=1)
df_test=pd.concat([df_test_16,df_test_18,df_test],axis=1)
columns=np.arange(0,42)
df_test=df_test.reindex(columns=columns)
df_test.head()

0                                1   2   3   \
0  38                          Private   6  36   
1  44   Self-employed-not incorporated  37  12   
2   2                  Not in universe   0   0   
3  35                          Private  29   3   
4  49                          Private   4  34   

                                     4   5                 6   \
0              1st 2nd 3rd or 4th grade   0   Not in universe   
1   Associates degree-occup /vocational   0   Not in universe   
2                              Children   0   Not in universe   
3                  High school graduate   0   Not in universe   
4                  High school graduate   0   Not in universe   

                                 7                              8   \
0   Married-civilian spouse present    Manufacturing-durable goods   
1   Married-civilian spouse present   Business and repair services   
2                     Never married    Not in universe or children   
3                          Divorced                 Transportation   
4                          Divorced                   Construction   

                                       9   ...              32  \
0   Machine operators assmblrs & inspctrs  ...          Mexico   
1                  Professional specialty  ...   United-States   
2                         Not in universe  ...   United-States   
3          Executive admin and managerial  ...   United-States   
4     Precision production craft & repair  ...   United-States   

               33              34                                    35 36  \
0          Mexico          Mexico   Foreign born- Not a citizen of U S   0   
1   United-States   United-States     Native- Born in the United States  0   
2   United-States   United-States     Native- Born in the United States  0   
3   United-States   United-States     Native- Born in the United States  2   
4   United-States   United-States     Native- Born in the United States  0   

                 37  38  39  40      41  
0   Not in universe   2  12  95  -50000  
1   Not in universe   2  26  95  -50000  
2   Not in universe   0   0  95  -50000  
3   Not in universe   2  52  94  -50000  
4   Not in universe   2  50  95  -50000  

[5 rows x 42 columns]

In [37]:
X_test=df_test.drop(columns=41,axis=1)

In [38]:
X_test_numeric=X_test.iloc[:,[0, 2, 3, 5, 16, 17, 18, 24, 30, 36, 38, 39, 40]]
X_test_numeric.head()

0   2   3   5    16  17      18       24  30  36  38  39  40
0  38   6  36   0  0.0   0     0.0  1032.38   4   0   2  12  95
1  44  37  12   0  0.0   0  2500.0  1462.33   1   0   2  26  95
2   2   0   0   0  0.0   0     0.0  1601.75   0   0   0   0  95
3  35  29   3   0  0.0   0     0.0  1866.88   5   2   2  52  94
4  49   4  34   0  0.0   0     0.0  1394.54   4   0   2  50  95

In [39]:
X_test_categorical=X_test.drop(columns=[0, 2, 3, 5, 16, 17, 18, 24, 30, 36, 38, 39, 40],axis=1)

In [40]:
X_test_categorical=X_test_categorical.apply(le.fit_transform)
X_test_categorical

1   4   6   7   8   9   10  11  12  13  ...  26  27  28  29  31  32  \
0       4   3   2   2  10   5   4   5   0   1  ...   8   9   1   3   4  25   
1       6   8   2   2   2  10   4   0   0   1  ...   8   9   1   3   4  39   
2       3  10   2   4  14   6   4   6   1   1  ...   8   9   1   3   0  39   
3       4  12   2   0  21   2   4   0   0   1  ...   5   6   2   1   4  39   
4       4  12   2   0   4   8   4   0   1   1  ...   8   9   1   3   4  39   
...    ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..   
99757   3  10   2   4  14   6   4   0   0   1  ...   8   9   1   3   0  39   
99758   4   1   2   5  10   5   2   0   1   0  ...   8   9   1   3   4  39   
99759   6   5   2   2   0   3   4   5   1   1  ...   5   6   2   1   4  25   
99760   4   9   2   2  15   2   4   0   0   1  ...   8   9   1   3   4  39   
99761   3   6   2   2  14   6   4   0   1   1  ...   5   6   2   1   4  39   

       33  34  35  37  
0      25  25   0   1  
1      39  39   4   1  
2      39  39   4   1  
3      39  39   4   1  
4      39  39   4   1  
...    ..  ..  ..  ..  
99757  39  39   4   1  
99758  39  39   4   1  
99759  25  25   1   1  
99760  39  39   4   1  
99761  39  39   4   1  

[99762 rows x 28 columns]

In [41]:
X_test=pd.concat([X_test_numeric,X_test_categorical],axis=1)

In [42]:
X_test=scaler.fit_transform(X_test)
X_test=pd.DataFrame(X_test)
X_test.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


0         1         2         3         4         5         6   \
0  0.150958 -0.517724  1.706065 -0.201558 -0.147614 -0.135278 -0.111594   
1  0.419615  1.201955  0.046259 -0.201558 -0.147614 -0.135278  1.384370   
2 -1.460982 -0.850565 -0.783644 -0.201558 -0.147614 -0.135278 -0.111594   
3  0.016630  0.758167 -0.576169 -0.201558 -0.147614 -0.135278 -0.111594   
4  0.643496 -0.628671  1.567748 -0.201558 -0.147614 -0.135278 -0.111594   

         7         8         9   ...        31        32        33        34  \
0 -0.710791  0.864454 -0.319722  ...  0.898859  0.874159 -0.549552  0.963901   
1 -0.278635 -0.404411 -0.319722  ...  0.898859  0.874159 -0.549552  0.963901   
2 -0.138500 -0.827365 -0.319722  ...  0.898859  0.874159 -0.549552  0.963901   
3  0.127990  1.287408  3.262496  ... -0.923544 -0.873143  1.077043 -0.929771   
4 -0.346773  0.864454 -0.319722  ...  0.898859  0.874159 -0.549552  0.963901   

         35        36        37        38        39        40  
0  0.585319 -1.306203 -1.322572 -1.864947 -3.240525  0.062262  
1  0.585319  0.369330  0.366716  0.269690  0.346029  0.062262  
2 -1.910358  0.369330  0.366716  0.269690  0.346029  0.062262  
3  0.585319  0.369330  0.366716  0.269690  0.346029  0.062262  
4  0.585319  0.369330  0.366716  0.269690  0.346029  0.062262  

[5 rows x 41 columns]

In [43]:
X_test=X_test.iloc[:,top_20.index]

In [44]:
X_test.shape

(99762, 20)

In [45]:
y_test=pd.DataFrame(df_test.iloc[:,41])
y_test

41
0      -50000
1      -50000
2      -50000
3      -50000
4      -50000
...       ...
99757  -50000
99758  -50000
99759  -50000
99760  -50000
99761  -50000

[99762 rows x 1 columns]

Evaluate the model on the test data.

In [46]:
y_test_pred=Tree.predict(X_test)
print("Accuracy on test data: ",accuracy_score(y_test,y_test_pred))

Accuracy on test data:  0.9521260600228544


In [47]:
from sklearn.metrics import confusion_matrix
print("Confusion matrix on test data: \n",confusion_matrix(y_test,y_test_pred))

Confusion matrix on test data: 
 [[ 3453  2733]
 [ 2043 91533]]


In [48]:
from sklearn.metrics import classification_report
print("Classification report on test data: \n",classification_report(y_test,y_test_pred))

Classification report on test data: 
               precision    recall  f1-score   support

     50000+.       0.63      0.56      0.59      6186
      -50000       0.97      0.98      0.97     93576

   micro avg       0.95      0.95      0.95     99762
   macro avg       0.80      0.77      0.78     99762
weighted avg       0.95      0.95      0.95     99762



The model is a bit over-trained: accuracy on train data is almost 100%, accuracy on test data is 95.21%. The model detected almost all of the cases in which income was below 50K and 63% of the cases in which income is above 50K. 

Oversampling the minority class increased the model's sensitivity to the minority class. However, because the oversampling was done through direct replication, the result is a slightly over-trained model. I think that if synthetic data was used, the result could have been more satisfactory, but it is difficult to generate realistic synthetic census data. 